File: clients/rust/src/instructions/update_oracle/update_oracle.rs
use crate::{
    derive_controller_authority_pda, generated::{instructions::UpdateOracleBuilder, types::FeedArgs},
};
use solana_instruction::Instruction;
use solana_pubkey::Pubkey;


pub fn create_update_oracle_instruction(
    controller: &Pubkey,
    authority: &Pubkey,
    oracle: &Pubkey,
    price_feed: &Pubkey,
    feed_args: Option<FeedArgs>,
    new_authority: Option<&Pubkey>,
) -> Instruction {
    let controller_authority = derive_controller_authority_pda(controller);
    let new_authority_pubkey = new_authority.map(|k| *k);
    let ixn = if let Some(feed_args) = feed_args {
        UpdateOracleBuilder::new()
            .controller(*controller)
            .controller_authority(controller_authority)
            .authority(*authority)
            .oracle(*oracle)
            .price_feed(*price_feed)
            .feed_args(feed_args)
            .new_authority(new_authority_pubkey)
            .instruction()
    } else {
        UpdateOracleBuilder::new()
            .controller(*controller)
            .controller_authority(controller_authority)
            .authority(*authority)
            .oracle(*oracle)
            .price_feed(*price_feed)
            .new_authority(new_authority_pubkey)
            .instruction()
    };
    ixn
}

File: clients/ts/src/instructions/updateOracle.ts
import { Address, TransactionSigner, OptionOrNullable } from "@solana/kit";
import { getUpdateOracleInstruction } from "../generated/instructions/updateOracle";
import { deriveControllerAuthorityPda } from "../pdas";
import { SVM_ALM_CONTROLLER_PROGRAM_ADDRESS } from "../generated";
import type { FeedArgs } from "../generated/types";

/**
 * Instruction generation for updating an oracle account
 */
export async function createUpdateOracleInstruction(
  controller: Address,
  authority: TransactionSigner,
  oracle: Address,
  priceFeed: Address,
  feedArgs: OptionOrNullable<FeedArgs>,
  newAuthority?: Address
) {
  const controllerAuthority = await deriveControllerAuthorityPda(controller);

  return getUpdateOracleInstruction({
    controller,
    controllerAuthority,
    authority,
    oracle,
    priceFeed,
    newAuthority,
    feedArgs,
  });
}

